### PR TITLE
Defines derivatives for remaining tgmath math functions.

### PR DESCRIPTION
**Pull request by vguerra**
* Inverse trigonometric functions.
* Exponents and logarithms.
* Hyperbolic functions.
* Error functions ( erf, erfc ).
* Free generic functions: sqrt, fma.

Partially resolves [TF-812](https://bugs.swift.org/browse/TF-812).

### CODE REVIEW

File: test/stdlib/tgmath.swift.gyb [hunk 1]
```diff
@@ -248,10 +248,26 @@ MathTests.test("${T}") {
 % for T in ['Float', 'Float80']:
 MathTests.test("gradient_${T}") {
   expectEqualWithTolerance(7.3890560989306502274, gradient(at: 2.0 as ${T}, in: exp), ulps:16)
+  expectEqualWithTolerance(2.772588722239781145, gradient(at: 2.0 as ${T}, in: exp2), ulps:16)
```
Comment (rxwei) [inline test/stdlib/tgmath.swift.gyb]:
```suggestion
  expectEqualWithTolerance(2.772588722239781145, gradient(at: 2.0 as ${T}, in: exp2), ulps: 16)
```
Same for all other occurrences.

File: stdlib/public/Platform/tgmath.swift.gyb [hunk 2]
```diff
@@ -102,11 +102,27 @@ func _vjpExp(_ x: ${T}) -> (${T}, (${T}) -> ${T}) {
...
 }
 
+@usableFromInline
+func _vjpLog10(_ x: ${T}) -> (${T}, (${T}) -> ${T}) {
+  return (log10(x), { v in v * ${T}(M_LOG10E) / x})
+}
+
+@usableFromInline
+func _vjpLog2(_ x: ${T}) -> (${T}, (${T}) -> ${T}) {
+    return (log2(x), { v in v / (${T}(M_LN2) * x)})
```
Comment (rxwei) [inline stdlib/public/Platform/tgmath.swift.gyb]: Indent by 2.

File: test/stdlib/tgmath.swift.gyb (hunk 1 above)
Comment (vguerra) [reply test/stdlib/tgmath.swift.gyb]: fixed on c92e34b5501ed136583f1e6e24f031f2dff53161

File: stdlib/public/Platform/tgmath.swift.gyb (hunk 2 above)
Comment (vguerra) [reply stdlib/public/Platform/tgmath.swift.gyb]: fixed on c92e34b5501ed136583f1e6e24f031f2dff53161

**Comment by vguerra** [comment]
I pushed 62c019b which implements derivates for `sqrt` and `fma`... thanks @rxwei for your explanation on the forums

File: stdlib/public/Platform/tgmath.swift.gyb [hunk 3]
```diff
@@ -20,11 +20,20 @@ public func fabs<T: FloatingPoint>(_ x: T) -> T {
...
+@differentiable(
+  vjp: _vjpSqrt
+  where T : Differentiable & FloatingPoint, T == T.TangentVector
+)
 public func sqrt<T: FloatingPoint>(_ x: T) -> T {
   return x.squareRoot()
 }
 
 @_transparent
+@differentiable(
```
Comment (rxwei) [inline stdlib/public/Platform/tgmath.swift.gyb]:
```suggestion
// SWIFT_ENABLE_TENSORFLOW
@differentiable(
```

File: stdlib/public/Platform/tgmath.swift.gyb [hunk 4]
```diff
@@ -20,11 +20,20 @@ public func fabs<T: FloatingPoint>(_ x: T) -> T {
...
```
Comment (rxwei) [inline stdlib/public/Platform/tgmath.swift.gyb]:
```suggestion
// SWIFT_ENABLE_TENSORFLOW
@differentiable(
```

**Review by rxwei — APPROVED** [review]

File: stdlib/public/Platform/tgmath.swift.gyb (hunk 4 above)
Comment (vguerra) [reply stdlib/public/Platform/tgmath.swift.gyb]: done in 27ca6a252ac2f46f4fe6c3f425dab2138fd439f3

File: stdlib/public/Platform/tgmath.swift.gyb (hunk 3 above)
Comment (vguerra) [reply stdlib/public/Platform/tgmath.swift.gyb]: done in 27ca6a252ac2f46f4fe6c3f425dab2138fd439f3

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

File: stdlib/public/Platform/tgmath.swift.gyb [hunk 5]
```diff
@@ -82,6 +93,25 @@ public func frexp<T: BinaryFloatingPoint>(_ x: T) -> (T, Int) {
...
+}
+
+@usableFromInline
+func _vjpFma<T: FloatingPoint & Differentiable> (
+  _ x: T,
+  _ y: T,
+  _ z: T
+) -> (T, (T) -> (T, T, T)) where T == T.TangentVector {
+  return (fma(x, y, z),
+    {v in return (v * y, v * x, v)})
```
Comment (rxwei) [inline stdlib/public/Platform/tgmath.swift.gyb]: * Move this to the end of the previous line.
* Add a space after each `{` before each `}`. Same for all other occurrences.
* Omit `return` when a closure has a single expression.

File: stdlib/public/Platform/tgmath.swift.gyb [hunk 6]
```diff
@@ -82,6 +93,25 @@ public func frexp<T: BinaryFloatingPoint>(_ x: T) -> (T, Int) {
   return (x.significand / 2, Int(x.exponent + 1))
 }
 
+// SWIFT_ENABLE_TENSORFLOW
+@usableFromInline
+func _vjpSqrt<T: FloatingPoint & Differentiable> (
+  _ x: T
+) -> (T, (T) -> T) where T == T.TangentVector {
+  let value = x.squareRoot()
+  return (value, {v in (1 / 2) * ( 1 / value) * v})
```
Comment (rxwei) [inline stdlib/public/Platform/tgmath.swift.gyb]:
```suggestion
  return (value, { v in v / (2 * value) })
```

File: stdlib/public/Platform/tgmath.swift.gyb [hunk 7]
```diff
@@ -102,11 +132,27 @@ func _vjpExp(_ x: ${T}) -> (${T}, (${T}) -> ${T}) {
...
+}
+
 @usableFromInline
 func _vjpLog(_ x: ${T}) -> (${T}, (${T}) -> ${T}) {
   return (log(x), { v in v / x })
 }
 
+@usableFromInline
+func _vjpLog10(_ x: ${T}) -> (${T}, (${T}) -> ${T}) {
+  return (log10(x), { v in v * ${T}(M_LOG10E) / x})
```
Comment (rxwei) [inline stdlib/public/Platform/tgmath.swift.gyb]:
```suggestion
  return (log10(x), { v in v * ${T}(M_LOG10E) / x })
```

File: stdlib/public/Platform/tgmath.swift.gyb [hunk 8]
```diff
@@ -102,11 +132,27 @@ func _vjpExp(_ x: ${T}) -> (${T}, (${T}) -> ${T}) {
...
+}
+
+@usableFromInline
+func _vjpLog2(_ x: ${T}) -> (${T}, (${T}) -> ${T}) {
+  return (log2(x), { v in v / (${T}(M_LN2) * x)})
```
Comment (rxwei) [inline stdlib/public/Platform/tgmath.swift.gyb]:
```suggestion
  return (log2(x), { v in v / (${T}(M_LN2) * x) })
```

File: stdlib/public/Platform/tgmath.swift.gyb [hunk 9]
```diff
@@ -122,6 +168,72 @@ func _vjpTan(_ x: ${T}) -> (${T}, (${T}) -> ${T}) {
...
+}
+
+@usableFromInline
+func _vjpExpm1(_ x: ${T}) -> (${T}, (${T}) -> ${T}) {
+  return (expm1(x), { v in exp(x) * v })
+}
+
+@usableFromInline
+func _vjpLog1p(_ x: ${T}) -> (${T}, (${T}) -> ${T}) {
+  return (log1p(x), { v in v / ( x + 1) })
```
Comment (rxwei) [inline stdlib/public/Platform/tgmath.swift.gyb]:
```suggestion
  return (log1p(x), { v in v / (x + 1) })
```

File: test/stdlib/tgmath.swift.gyb [hunk 10]
```diff
@@ -247,11 +247,32 @@ MathTests.test("${T}") {
...
+  expectEqualWithTolerance(3.7621956910836314597, gradient(at: 2.0 as ${T}, in: sinh), ulps: 16)
+  expectEqualWithTolerance(3.6268604078470187677, gradient(at: 2.0 as ${T}, in: cosh), ulps: 16)
+  expectEqualWithTolerance(0.07065082485316446565, gradient(at: 2.0 as ${T}, in: tanh), ulps: 16)
+  expectEqualWithTolerance(0.44721359549995793928, gradient(at: 2.0 as ${T}, in: asinh), ulps: 16)
+  expectEqualWithTolerance(0.5773502691896257645, gradient(at: 2.0 as ${T}, in: acosh), ulps: 16)
+  expectEqualWithTolerance(1.3333333333333333334, gradient(at: 0.5 as ${T}, in: atanh), ulps: 16)
+  expectEqualWithTolerance(0.020666985354092053575, gradient(at: 2.0 as ${T}, in: erf), ulps: 16)
+  expectEqualWithTolerance(-0.020666985354092053575, gradient(at: 2.0 as ${T}, in: erfc), ulps: 16)
+  expectEqualWithTolerance(0.35355339059327376222, gradient(at: 2.0 as ${T}, in: {x in sqrt(x)}), ulps: 16)
+  let fma_grad = gradient(at: 4.0 as ${T}, 5.0 as ${T}, 6.0 as ${T}, in: {x, y, z in fma(x, y, z)})
```
Comment (rxwei) [inline test/stdlib/tgmath.swift.gyb]:
```suggestion
  let fmaGrad = gradient(at: 4.0 as ${T}, 5.0 as ${T}, 6.0 as ${T}, in: { x, y, z in fma(x, y, z) })
```

File: test/stdlib/tgmath.swift.gyb [hunk 11]
```diff
@@ -247,11 +247,32 @@ MathTests.test("${T}") {
...
+  expectEqualWithTolerance(0.8, gradient(at: 0.5 as ${T}, in: atan), ulps: 16)
+  expectEqualWithTolerance(3.7621956910836314597, gradient(at: 2.0 as ${T}, in: sinh), ulps: 16)
+  expectEqualWithTolerance(3.6268604078470187677, gradient(at: 2.0 as ${T}, in: cosh), ulps: 16)
+  expectEqualWithTolerance(0.07065082485316446565, gradient(at: 2.0 as ${T}, in: tanh), ulps: 16)
+  expectEqualWithTolerance(0.44721359549995793928, gradient(at: 2.0 as ${T}, in: asinh), ulps: 16)
+  expectEqualWithTolerance(0.5773502691896257645, gradient(at: 2.0 as ${T}, in: acosh), ulps: 16)
+  expectEqualWithTolerance(1.3333333333333333334, gradient(at: 0.5 as ${T}, in: atanh), ulps: 16)
+  expectEqualWithTolerance(0.020666985354092053575, gradient(at: 2.0 as ${T}, in: erf), ulps: 16)
+  expectEqualWithTolerance(-0.020666985354092053575, gradient(at: 2.0 as ${T}, in: erfc), ulps: 16)
+  expectEqualWithTolerance(0.35355339059327376222, gradient(at: 2.0 as ${T}, in: {x in sqrt(x)}), ulps: 16)
```
Comment (rxwei) [inline test/stdlib/tgmath.swift.gyb]:
```suggestion
  expectEqualWithTolerance(0.35355339059327376222, gradient(at: 2.0 as ${T}, in: { sqrt($0) }), ulps: 16)
```

File: test/stdlib/tgmath.swift.gyb (hunk 10 above)
Comment (rxwei) [inline test/stdlib/tgmath.swift.gyb]: Always use camel case.

```suggestion
  let fmaGrad = gradient(at: 4.0 as ${T}, 5.0 as ${T}, 6.0 as ${T}, in: {x, y, z in fma(x, y, z)})
```

**Review by rxwei — CHANGES_REQUESTED** [review]

**Comment by rxwei** [comment]
@stephentyrone Could you help review the usage of math constants in these derivatives?

File: stdlib/public/Platform/tgmath.swift.gyb (hunk 5 above)
Comment (vguerra) [reply stdlib/public/Platform/tgmath.swift.gyb]: done in 6b1ebf264b6801e5becf1b9f5d6206ce627af1f6

File: stdlib/public/Platform/tgmath.swift.gyb (hunk 6 above)
Comment (vguerra) [reply stdlib/public/Platform/tgmath.swift.gyb]: done in 6b1ebf264b6801e5becf1b9f5d6206ce627af1f6

File: stdlib/public/Platform/tgmath.swift.gyb (hunk 7 above)
Comment (vguerra) [reply stdlib/public/Platform/tgmath.swift.gyb]: done in 6b1ebf264b6801e5becf1b9f5d6206ce627af1f6

File: stdlib/public/Platform/tgmath.swift.gyb (hunk 8 above)
Comment (vguerra) [reply stdlib/public/Platform/tgmath.swift.gyb]: done in 6b1ebf264b6801e5becf1b9f5d6206ce627af1f6

File: stdlib/public/Platform/tgmath.swift.gyb (hunk 9 above)
Comment (vguerra) [reply stdlib/public/Platform/tgmath.swift.gyb]: done in 6b1ebf264b6801e5becf1b9f5d6206ce627af1f6

File: test/stdlib/tgmath.swift.gyb (hunk 10 above)
Comment (vguerra) [reply test/stdlib/tgmath.swift.gyb]: done in 6b1ebf264b6801e5becf1b9f5d6206ce627af1f6

Comment (vguerra) [reply test/stdlib/tgmath.swift.gyb]: done in 6b1ebf264b6801e5becf1b9f5d6206ce627af1f6

File: test/stdlib/tgmath.swift.gyb (hunk 11 above)
Comment (vguerra) [reply test/stdlib/tgmath.swift.gyb]: done in 6b1ebf264b6801e5becf1b9f5d6206ce627af1f6

**Review by rxwei — APPROVED** [review]

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Comment by rxwei** [comment]
@swift-ci please test tensorflow

**Pull request merged**